napari/_vispy/vispy_image_layer.py

```diff
@@ -143,22 +144,42 @@ def _on_threshold_change(self, event=None):
             self.node.threshold = float(self.layer.attenuation)
 
     def _on_scale_change(self, event=None):
-        self.scale = [
-            self.layer.scale[d] * self.layer._scale_view[d]
-            for d in self.layer.dims.displayed[::-1]
+        name = 'scale_vispy_image_layer'
+        new_scale_values = [
+            self.layer.scale[d] for d in self.layer.dims.displayed[::-1]
         ]
+        new_transform = Scale(new_scale_values, name=name)
+        try:
+            self.layer.transforms[name]
+        except KeyError:
+            pass
+        else:
+            self.layer.transforms.remove(self.layer.transforms[name])
+        finally:
+            self.layer.transforms.append(new_transform)
+
         if self.layer.is_pyramid:
             self.layer.top_left = self.find_top_left()
         self.layer.position = self._transform_position(self._position)
 
     def _on_translate_change(self, event=None):
-        self.translate = [
+        name = 'translate_vispy_base_layer'
+        new_translate_values = [
             self.layer.translate[d]
             + self.layer._translate_view[d]
             + self.layer.translate_grid[d]
             for d in self.layer.dims.displayed[::-1]
         ]
-        self.layer.position = self._transform_position(self._position)
+        new_transform = Translate(new_translate_values, name=name)
+        try:
+            self.layer.transforms[name]
+        except KeyError:
+            pass
+        else:
+            self.layer.transforms.remove(self.layer.transforms[name])
+        finally:
+            self.layer.transforms.append(new_transform)
+            self.layer.position = self._transform_position(self._position)
```

Comment:
Is this identical to the base layer? In which case it shouldn't be needed?

Reply:
Almost, but not quite identical. You make a good point, so I have consolidated `_on_translate_change()`.

The not-identical parts are to do with handling image pyramids. 
The only difference between the `_on_scale_change()` method in `vispy_base_layer.py` vs `vispy_image_layer.py` is this if block:
```python
        if self.layer.is_pyramid:
            self.layer.top_left = self.find_top_left()
```

The `layer.is_pyramid` attribute only exists for `Image` and `Labels` layer types.

**This is what I'm going to do:
Add the `is_pyramid` attribute to the base layer (default value False). Then I will add those two lines into the vispy base layer `_on_scale_change()` method and remove the two overriding methods from the vispy image layer & the vispy labels layer.**

@sofroniewn if you have an objection to this plan please let me know, if not I think this is the cleanest way forward.

Reply:
I was wrong, [the is_pyramid attribute *does* exist in the base layer](https://github.com/napari/napari/blob/323c24c1d06378dcf2f654033c90bcdcac8f1a58/napari/layers/base/base.py#L146), I just missed that line because it wasn't in the docstring. 

Too easy, I've consolidated the `_on_scale_change()` method now.